llvm/lib/Target/Mips/MCTargetDesc/MipsABIInfo.h

```diff
@@ -26,19 +26,46 @@ class MipsABIInfo {
 public:
   enum class ABI { Unknown, O32, N32, N64 };
 
+  class TemporalABILayout {
+    uint32_t CapSize;
+
+  public:
+    int64_t GetThreadLocalOffset_CSP() const { return CapSize * 4; }
+    int64_t GetThreadLocalOffset_CUSP() const { return CapSize * 5; }
+    int64_t GetThreadLocalOffset_CDS() const { return CapSize * 6; }
+    int64_t GetThreadLocalOffset_CDL() const { return CapSize * 7; }
+    int64_t GetThreadLocalOffset_CGP() const { return CapSize * 8; }
+
+    int64_t GetStackPointerOffset_Next() const { return CapSize * -1; }
+    int64_t GetStackPointerOffset_Prev() const { return CapSize * -2; }
+    int64_t GetStackPointerSize_Prev() const { return CapSize * 2; }
+    int64_t GetStackPointerAlign_Prev() const { return CapSize; }
```

Comment:
These should all be offsets, so we don't need to pass the capSize down via your ugly `CPU.endsWith("128")` hack. Then you can use the TRI available in the frame lowering code to scale it appropriately. And at that point this just becomes a list of constants.

Reply:
Done